flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/catalog/CatalogSchemaModel.java

```diff
@@ -85,4 +109,34 @@ private static RelDataType schemaToRelDataType(
                         .collect(Collectors.toList());
         return typeFactory.buildRelNodeRowType(fieldNames, fieldTypes);
     }
+
+    private ModelProvider createModelProvider(
+            FlinkContext context, ContextResolvedModel catalogModel) {
+
+        final Optional<ModelProviderFactory> factoryFromCatalog =
+                catalogModel
+                        .getCatalog()
+                        .flatMap(Catalog::getFactory)
+                        .map(
+                                f ->
+                                        f instanceof ModelProviderFactory
+                                                ? (ModelProviderFactory) f
+                                                : null);
+
+        final Optional<ModelProviderFactory> factoryFromModule =
+                context.getModuleManager().getFactory(Module::getModelProviderFactory);
+
+        // Since the catalog is more specific, we give it precedence over a factory provided by any
+        // modules.
+        final ModelProviderFactory factory =
+                firstPresent(factoryFromCatalog, factoryFromModule).orElse(null);
+
+        return FactoryUtil.createModelProvider(
```

Comment:
The `PredictRuntimeProvider` and `AsyncPredictRuntimeProvider` is a problem here. We don't have access to the async config. The provider is choosed based on `provider` option only

Reply:
I put `createAsyncPredictFunction` inside `PredictRuntimeProvider` and removed `AsyncPredictRuntimeProvider`. @fsk119

Reply:
I think the actual provider should implement both interfaces and then let the framework to determine which actual function is used in the later stage. 

If the model developer prefer to split the implementation, I think we can convert SqlMLPredictTableFunction in total rather than convert ModelCall. 

```
 public SqlRexConvertlet get(SqlCall call) {
        ...
        if (isMLPredict(call)) {
            return this::convertSqlMLPredictFunction;
        }

        return StandardConvertletTable.INSTANCE.get(call);
    }

```